### PR TITLE
[MRG] Add "precomputed" option for dtw functions

### PR DESCRIPTION
**Pull request by hichamjanati**
This PR adds a "precomputed" option for the `dist` arg when computing`dtw`. The user can choose whether to provide `x` and `y` or a precomputed cost matrix through a new `precomputed_cost` arg. 

- [x] Add the new args + adapt computation of `region` + `cost_mat`.
- [x] Add tests with precomputed cost matrices (test against square for e.g)
- [x] A large portion of the code in the different dtw functions is redundant: shorten it

### CODE REVIEW

**Comment by hichamjanati** [comment]
All the dtw functions have the same structure: 
1. check inputs
2. compute_region 
3. input + region ->  cost_mat
4. compute acc_cost_mat

I'm thinking we could have a single dtw function (above) and a new compute_region func that contains the if...else on `method` instead of having multiple dtw_* functions.
For this PR I would however keep the dtw_ functions and add a deprecating warning.
 
WDYT @johannfaouzi  ?

**Comment by johannfaouzi** [comment]
A bit off topic: [it is possible to enable automatic documentation building for pull requests with Read the Docs](https://docs.readthedocs.io/en/stable/guides/autobuild-docs-for-pull-requests.html). Do you think that it is worth it? I think it could be nice to check the documentation on RTD instead of building it ourselves for each PR.

**Comment by hichamjanati** [comment]
I included all your remarks. The coverage decreased because of the `deprecated` module (only tested for functions and not objects / methods)
I think we're good to go !

File: pyts/metrics/dtw.py [hunk 1]
```diff
@@ -239,25 +277,32 @@ def _return_results(dtw_dist, cost_mat, acc_cost_mat,
     else:
         return res
 
-
-def dtw_classic(x, y, dist='square', return_cost=False,
-                return_accumulated=False, return_path=False):
+@deprecated("``dtw_classic`` is deprecated in v0.9.0 and will be removed in "
```
[the remark this answers is not in the document]
Comment (hichamjanati) [reply pyts/metrics/dtw.py]: Yes you're right it should be 0.10 now. I think we can remove the functions in 0.11, it is not a major change as `dtw` does the exact same thing.

File: pyts/metrics/dtw.py [hunk 2]
```diff
@@ -306,21 +351,25 @@ def dtw_classic(x, y, dist='square', return_cost=False,
...
 
     Parameters
     ----------
     x : array-like, shape = (n_timestamps_1,)
-        First array.
+        First array. ``dist == 'precomputed'``.
 
     y : array-like, shape = (n_timestamps_2,)
-        Second array
+        Second array. ``dist == 'precomputed'``.
```
[the remark this answers is not in the document]
Comment (hichamjanati) [reply pyts/metrics/dtw.py]: Why you should always be careful with CTRL-F ... :)

File: pyts/metrics/dtw.py [hunk 3]
```diff
@@ -766,9 +944,9 @@ def dtw_multiscale(x, y, dist='square', resolution=2, radius=0,
         First array.
 
     y : array-like, shape = (n_timestamps_2,)
-        Second array
+        Second array.
```
[the remark this answers is not in the document]
Comment (hichamjanati) [reply pyts/metrics/dtw.py]: Yes you're right, an error is raised with `multiscale` and `fast`. It's added in the doc now; and removed the corresponding params for `dtw_fast` and `dtw_multiscale`.

**Comment by johannfaouzi** [comment]
Thanks a lot!

**Pull request merged**